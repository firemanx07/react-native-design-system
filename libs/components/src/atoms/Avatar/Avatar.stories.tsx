import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { StorybookScreen } from '../../storybook';
import {
  Avatar as AvatarBase,
  PropsType as AvatarProps,
  AvatarSize,
} from './Avatar';

const Story: ComponentMeta<typeof AvatarBase> = {
  component: AvatarBase,
  title: 'Atoms/Avatar',
};
export default Story;

const Template: ComponentStory<typeof AvatarBase> = (args: AvatarProps) => (
  <StorybookScreen title="Avatar:">
    <AvatarBase {...args} />
  </StorybookScreen>
);

const parameters = {
  controls: {
    include: ['uri', 'size'],
  },
};

const defaultArgs: AvatarProps = {
  size: AvatarSize.S,
};

export const Avatar = Template.bind({});
Avatar.parameters = parameters;
Avatar.args = {
  uri: 'https://picsum.photos/120/120',
  ...defaultArgs,
};

export const AvatarPlaceholder = Template.bind({});
AvatarPlaceholder.parameters = parameters;
AvatarPlaceholder.args = {
  ...defaultArgs,
};
