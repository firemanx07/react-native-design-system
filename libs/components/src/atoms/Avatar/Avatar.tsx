import { styled } from '@proxym/themes';
import React, { memo, useMemo } from 'react';
import { ImageSourcePropType } from 'react-native';

import { TestIDType } from '../../types';

export enum AvatarSize {
  XS = 20,
  S = 30,
  M = 36,
  L = 44,
  XL = 120,
}

export type PropsType = {
  uri?: string;
  size?: AvatarSize;
} & TestIDType;

export const Avatar = ({ uri, size = AvatarSize.S, testID }: PropsType) => {
  const imageSource: ImageSourcePropType = useMemo(() => {
    if (uri) {
      return { uri };
    }

    return require('../../assets/images/avatar-placeholder.png');
  }, [size, uri]);

  return (
    <Container size={size} testID={testID}>
      <AvatarImage size={size} source={imageSource} />
    </Container>
  );
};

type StyledCommonType = {
  size: AvatarSize;
};

const Container = styled.View<StyledCommonType>`
  height: ${({ size }) => size}px;
  width: ${({ size }) => size}px;
  border-radius: ${({ size }) => size / 2}px;
  overflow: hidden;
`;

const AvatarImage = styled.Image<StyledCommonType>`
  height: ${({ size }) => size}px;
  width: ${({ size }) => size}px;
`;

export default memo(Avatar);
