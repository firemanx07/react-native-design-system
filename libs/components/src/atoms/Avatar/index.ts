export { default as Avatar, AvatarSize } from './Avatar';
