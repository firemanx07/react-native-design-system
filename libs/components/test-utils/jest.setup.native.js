import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();

console.warn = () => {};

Animated.timing = () => ({
  start: () => jest.fn(),
});
