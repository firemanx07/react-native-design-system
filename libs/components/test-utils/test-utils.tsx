import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
