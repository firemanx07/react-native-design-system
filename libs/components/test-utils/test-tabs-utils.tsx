import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
