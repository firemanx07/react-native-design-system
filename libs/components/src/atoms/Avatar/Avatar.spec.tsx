import { testAllStories } from '../../../test-utils';
import * as stories from './Avatar.stories';

describe('Avatar tests', () => {
  testAllStories(stories);
});
