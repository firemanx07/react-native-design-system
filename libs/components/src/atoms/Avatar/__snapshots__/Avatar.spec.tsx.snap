// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar tests Avatar snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F2F3F4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
    ]
  }
>
  <Text
    fontFamily="Muli-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#222222",
          "fontFamily": "Muli-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    Avatar:
  </Text>
  <View
    size={30}
    style={
      [
        {
          "borderRadius": 15,
          "height": 30,
          "overflow": "hidden",
          "width": 30,
        },
      ]
    }
  >
    <Image
      size={30}
      source={
        {
          "uri": "https://picsum.photos/120/120",
        }
      }
      style={
        [
          {
            "height": 30,
            "width": 30,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`Avatar tests AvatarPlaceholder snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F2F3F4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
    ]
  }
>
  <Text
    fontFamily="Muli-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#222222",
          "fontFamily": "Muli-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    Avatar:
  </Text>
  <View
    size={30}
    style={
      [
        {
          "borderRadius": 15,
          "height": 30,
          "overflow": "hidden",
          "width": 30,
        },
      ]
    }
  >
    <Image
      size={30}
      source={
        {
          "uri": "",
        }
      }
      style={
        [
          {
            "height": 30,
            "width": 30,
          },
        ]
      }
    />
  </View>
</View>
`;
